File: java/src/tarehart/rlbot/steps/demolition/DemolishEnemyStep.java
package tarehart.rlbot.steps.demolition;

import tarehart.rlbot.AgentInput;
import tarehart.rlbot.AgentOutput;
import tarehart.rlbot.carpredict.AccelerationModel;
import tarehart.rlbot.carpredict.CarInterceptPlanner;
import tarehart.rlbot.carpredict.CarPath;
import tarehart.rlbot.carpredict.CarPredictor;
import tarehart.rlbot.input.CarData;
import tarehart.rlbot.math.SpaceTime;
import tarehart.rlbot.physics.DistancePlot;
import tarehart.rlbot.planning.*;
import tarehart.rlbot.steps.Step;
import tarehart.rlbot.time.Duration;

import java.awt.*;
import java.util.Optional;

public class DemolishEnemyStep implements Step {

    private boolean enemyHadWheelContact;

    public Optional<AgentOutput> getOutput(AgentInput input) {

        Optional<CarData> enemyCarOption = input.getEnemyCarData();

        CarData car = input.getMyCarData();
        if (!enemyCarOption.isPresent() || enemyCarOption.map(CarData::isDemolished).orElse(false) || (car.getBoost() == 0 && !car.isSupersonic())) {
            return Optional.empty();
        }

        CarData enemyCar = enemyCarOption.get();

        CarPath path = CarPredictor.predictCarMotion(enemyCar, Duration.Companion.ofSeconds(4));

        DistancePlot distancePlot = AccelerationModel.INSTANCE.simulateAcceleration(car, Duration.Companion.ofSeconds(4), car.getBoost());

        Optional<SpaceTime> carInterceptOption = CarInterceptPlanner.getCarIntercept(car, path, distancePlot);

        if (carInterceptOption.isPresent()) {
            SpaceTime carIntercept = carInterceptOption.get();

            // TODO: deal with cases where car is driving toward arena boundary

//        if (ArenaModel.isInBoundsBall(path.getSlices().get(path.getSlices().size() - 1).space)) {
//            BotLog.println("Whoops", car.playerIndex);
//        }

            AgentOutput steering = SteerUtil.steerTowardGroundPosition(car, carIntercept.getSpace());

            double secondsTillContact = Duration.Companion.between(car.getTime(), carIntercept.getTime()).getSeconds();

            if (secondsTillContact < .5 && !enemyCar.getHasWheelContact() && (enemyHadWheelContact || enemyCar.getPosition().getZ() - car.getPosition().getZ() > 1)) {
                steering.withJump();
                if (!car.getHasWheelContact()) {
                    steering.withSteer(0); // Avoid dodging accidentally.
                }
            }

            enemyHadWheelContact = enemyCar.getHasWheelContact();

            return Optional.of(steering);
        }

        return Optional.of(SteerUtil.steerTowardGroundPosition(car, enemyCar.getPosition()));
    }

    @Override
    public boolean canInterrupt() {
        return true;
    }

    @Override
    public String getSituation() {
        return "Demolishing enemy";
    }

    @Override
    public void drawDebugInfo(Graphics2D graphics) {
        // Draw nothing.
    }
}

File: java/src/tarehart/rlbot/steps/demolition/DemolishEnemyStep.kt
package tarehart.rlbot.steps.demolition

import tarehart.rlbot.AgentInput
import tarehart.rlbot.AgentOutput
import tarehart.rlbot.carpredict.AccelerationModel
import tarehart.rlbot.carpredict.CarInterceptPlanner
import tarehart.rlbot.carpredict.CarPredictor
import tarehart.rlbot.planning.*
import tarehart.rlbot.steps.Step
import tarehart.rlbot.time.Duration

import java.awt.*
import java.util.Optional

class DemolishEnemyStep : Step {

    private var enemyHadWheelContact: Boolean = false
    private var hasDoubleJumped: Boolean = false
    private var enemyIndex: Int? = null

    override val situation = "Demolishing enemy"

    override fun getOutput(input: AgentInput): AgentOutput? {

        val car = input.myCarData
        val oppositeTeam = input.getTeamRoster(input.team.opposite())
        val enemyCar = enemyIndex?.let { enemyInd -> oppositeTeam.first { it.playerIndex == enemyInd } } ?:
                oppositeTeam.minBy { car.position.distance(it.position) }

        if (enemyCar == null || enemyCar.isDemolished || car.boost == 0.0 && !car.isSupersonic) {
            return null
        }

        if (enemyCar.position.distance(car.position) < 30) {
            enemyIndex = enemyCar.playerIndex // Commit to demolishing this particular enemy
        }

        val path = CarPredictor.predictCarMotion(enemyCar, Duration.ofSeconds(4.0))

        val distancePlot = AccelerationModel.simulateAcceleration(car, Duration.ofSeconds(4.0), car.boost)

        val carIntercept = CarInterceptPlanner.getCarIntercept(car, path, distancePlot)

        carIntercept?.let {

            // TODO: deal with cases where car is driving toward arena boundary

            //        if (ArenaModel.isInBoundsBall(slices.getSlices().get(slices.getSlices().size() - 1).space)) {
            //            BotLog.println("Whoops", car.playerIndex);
            //        }

            val steering = SteerUtil.steerTowardGroundPosition(car, it.space)

            val secondsTillContact = Duration.between(car.time, it.time).seconds

            val heightAtIntercept = it.space.z
            val needsDoubleJump = heightAtIntercept > 5

            val heightDifference = heightAtIntercept - car.position.z
            if (secondsTillContact < .8 && !enemyCar.hasWheelContact && (enemyHadWheelContact || heightDifference > 1)) {

                if (needsDoubleJump && !hasDoubleJumped && !car.hasWheelContact) {
                    // Let go of A for a moment
                    hasDoubleJumped = true
                } else {
                    steering.withJump()
                }


                if (!car.hasWheelContact) {
                    steering.withSteer(0.0) // Avoid dodging accidentally.
                }
            }

            enemyHadWheelContact = enemyCar.hasWheelContact

            return steering
        }

        return SteerUtil.steerTowardGroundPosition(car, input.boostData, enemyCar.position.flatten())
    }

    override fun canInterrupt(): Boolean {
        return true
    }

    override fun drawDebugInfo(graphics: Graphics2D) {
        // Draw nothing.
    }
}
